modulus/utils/neighbor_list.py

```diff
@@ -0,0 +1,135 @@
+    index = int(0)
+    result_count_tid = int(0)
+
+    while wp.hash_grid_query_next(query, index):
+        neighbor = points[index]
+
+        # compute distance to neighbor point
+        dist = wp.length(qp - neighbor)
+        if dist <= radius:
+            result_count_tid += 1
+
+    result_count[tid] = result_count_tid
+
+
+@wp.kernel
+def radius_search_query(
+    hashgrid: wp.uint64,
+    points: wp.array(dtype=wp.vec3),
+    queries: wp.array(dtype=wp.vec3),
+    result_offset: wp.array(dtype=wp.int32),
+    result_point_idx: wp.array(dtype=wp.int32),
+    result_point_dist: wp.array(dtype=wp.float32),
+    radius: wp.float32,
+):
+    tid = wp.tid()
+
+    # create grid query around point
+    qp = queries[tid]
+    query = wp.hash_grid_query(hashgrid, qp, radius)
+    index = int(0)
+    result_count = int(0)
+    offset_tid = result_offset[tid]
+
+    while wp.hash_grid_query_next(query, index):
+        neighbor = points[index]
+
+        # compute distance to neighbor point
+        dist = wp.length(qp - neighbor)
+        if dist <= radius:
+            result_point_idx[offset_tid + result_count] = index
+            result_point_dist[offset_tid + result_count] = dist
+            result_count += 1
+
+
+def radius_search(
+    points: wp.array(dtype=wp.vec3),
+    queries: wp.array(dtype=wp.vec3),
+    radius: float,
+    grid_dim: Union[int, Tuple[int, int, int]] = (128, 128, 128),
+    device: str = "cuda",
+):
+    # convert grid_dim to Tuple if it is int
+    if isinstance(grid_dim, int):
+        grid_dim = (grid_dim, grid_dim, grid_dim)
+
+    result_count = wp.zeros(shape=queries.shape, dtype=wp.int32)
+    grid = wp.HashGrid(
+        dim_x=grid_dim[0],
+        dim_y=grid_dim[1],
+        dim_z=grid_dim[2],
+        device=device,
+    )
+    grid.build(points=points, radius=2 * radius)
+
+    wp.launch(
+        kernel=radius_search_count,
+        dim=len(queries),
+        inputs=[grid.id, points, queries, result_count, radius],
+        device=device,
+    )
+
+    torch_offset = torch.zeros(len(result_count) + 1, device=device, dtype=torch.int32)
+    result_count_torch = wp.to_torch(result_count)
+    torch.cumsum(result_count_torch, dim=0, out=torch_offset[1:])
+    total_count = torch_offset[-1].item()
+    if not total_count < 2**31 - 1:
+        raise RuntimeError(
+            f"Total result count is too large: {total_count} > 2**31 - 1"
+        )
+    offset = wp.from_torch(torch_offset)
+
+    result_point_idx = wp.zeros(shape=(total_count,), dtype=wp.int32)
+    result_point_dist = wp.zeros(shape=(total_count,), dtype=wp.float32)
+
+    wp.launch(
+        kernel=radius_search_query,
+        dim=len(queries),
+        inputs=[
+            grid.id,
+            points,
+            queries,
+            offset,
+            result_point_idx,
+            result_point_dist,
+            radius,
+        ],
+        device=device,
+    )
+
+    return (result_point_idx, result_point_dist, torch_offset)
```

Comment:
I think this gives a mix of warp and pytorch arrays. Might want to stick with one.

Reply:
Fixed